File: 2ndSemester/kursprogramowania/lista3/cpp/src/rectangle.cpp
#include <rectangle.hpp>

#include <cmath>
#include <string>

Rectangle::Rectangle(double side1, double side2) noexcept(false)
{
    if (side1 <= 0 || side2 <= 0)
        throw std::invalid_argument(std::string(__FILE__) + ":" + std::to_string(__LINE__) + " [" + __func__ + "] " + "side lengths should be positive, got " + std::to_string(side1) + " and " + std::to_string(side2));

    this->side1 = side1;
    this->side2 = side2;
}

double Rectangle::area() const noexcept(true) 
{
    return this->side1 * this->side2;
}

double Rectangle::perimeter() const noexcept(true)
{
    return 2 * this->side1 + 2 * this->side2;
}

const std::string Rectangle::name() const noexcept(true) 
{
    return "Rectangle";
}

File: 2ndSemester/kursprogramowania/lista3/java/Rectangle.java
/**
 * Rectangle
 */
public final class Rectangle extends Quadrilateral 
{
    private final double side1;
    private final double side2;
    
    public Rectangle(final double side1, final double side2)
    {
        if (side1 <= 0 || side2 <= 0) 
        {
            throw new IllegalArgumentException("Side lengths should be positive, got " + side1 + " and " + side2);    
        }
        this.side1 = side1;
        this.side2 = side2;
    }

    @Override
    public double area() 
    {
        return side1 * side2;
    } 

    @Override
    public double perimeter()
    {
        return 2 * side1 + 2 * side2;
    }

    @Override
    public String name()
    {
        return "Rectangle";
    }
}
